numpy/core/src/multiarray/alloc.c

```diff
@@ -113,8 +113,11 @@ npy_free_cache(void * p, npy_uintp sz)
 NPY_NO_EXPORT void *
 npy_alloc_cache_dim(npy_uintp sz)
 {
-    /* dims + strides */
-    if (NPY_UNLIKELY(sz < 2)) {
+    /*
+     * make sure any temporary allocation can be used for array metadata which
+     * uses one memory block for both dimensions and strides
+     */
```

Comment:
hm right it there was only one path, it was premature optimization then. I probably planned to do this change a lot sooner ;)
for something to miss one already has to deal with 3d arrays. these are pretty unlikely to be so small that the cache is relevant.
The cache really does not need to be perfect. It just needs to cover the extremely common alloc + immediate dealloc loop which it does.